### ISSUE TITLE
InferParameterBindingInfoConvention incorrectly treats CancellationToken as a body binding parameter

### ISSUE THREAD

**Issue by blushingpenguin** [aspnet/AspNetCore#13045]
### Describe the bug
```csharp
[ApiController]
class TestController : Controller
{
    [HttpPost("/Test")]
    public async Task DummyAction(CancellationToken ct, [FromBody] TestObj testObj)
    {
        // ....
```
leads to something along the lines of:
```
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> 
System.InvalidOperationException: Action 'TestController.DummyAction' has more than one parameter that was specified or inferred as bound from request body. Only one parameter per action may be bound from body. Inspect the following parameters, and use 'FromQueryAttribute' to specify bound from query, 'FromRouteAttribute' to specify bound from route, and 'FromBodyAttribute' for parameters to be bound from body:
CancellationToken ct
TestObj testObj
at Microsoft.AspNetCore.Mvc.ApplicationModels.InferParameterBindingInfoConvention.InferParameterBindingSources(ActionModel action)
```

in InferBindingSourceForParameter, 

### To Reproduce
1. Using v2.2.6 of ASP.NET Core
2. Run nswag on a webapi with the code above
3. With default settings

### Expected behavior
The binding source should be inferred as BindingSource.Special which is consistent with 
```modelMetadataDetailsProviders.Add(new BindingSourceMetadataProvider(typeof(CancellationToken), BindingSource.Special));``` in ```MvcCoreMvcOptionsSetup.ConfigureAdditionalModelMetadataDetailsProviders```
Changing InferBindingSourceForParameter to:

```csharp
internal BindingSource InferBindingSourceForParameter(ParameterModel parameter)
{
    if (parameter.ParameterType == typeof(CancellationToken))
    {
        return BindingSource.Special;
    }

    if (IsComplexTypeParameter(parameter))
    {
        return BindingSource.Body;
    }

    if (ParameterExistsInAnyRoute(parameter.Action, parameter.ParameterName))
    {
        return BindingSource.Path;
    }

    return BindingSource.Query;
}
```

prevents the error message as there is no longer a source conflict.

It could also be that the actual bug here is that CancellationTokens should not be added to ActionModel.Parameters at all.  Since they can only ever be local to the server and are always bound by asp.net to a token for the current request then I would consider them part of the implementation of an API rather than part of the public interface.  This would also solve the tangentially related:

https://github.com/aspnet/AspNetCore/issues/7865

and this issue in nswag:
https://github.com/RicoSuter/NSwag/issues/2347

and would have skipped some historical issues in swashbuckle:
https://github.com/domaindrivendev/Swashbuckle/issues/306
https://github.com/domaindrivendev/Swashbuckle/issues/744

so I think there's a strong case for just ignoring them.

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us, @blushingpenguin.
@pranavkm can you please look into this? Thanks!

**Comment by mkArtakMSFT** [comment]
We need to make sure this works properly in 3.0.

**Comment by pranavkm** [comment]
@blushingpenguin it looks like your app is using a 2.0 compatibility version:

> Compatibility switch AllowInferringBindingSourceForCollectionTypesAsFromQuery in type ApiBehaviorOptions is using compatibility value True for version Version_2_0

We made a couple of fixes in 2.2 to address some of this, and I think `CancellationToken` was one of them. However, it requires a newer compatibility version. Could you try updating your application and let me know if that addresses your issue?

**Comment by blushingpenguin** [comment]
Yes, that solves all my problems! It appears that with the compatibility version set to >= 2.1 cancellation tokens are filtered out from the list of parameters (I checked this with a generated asp.net core app and some code to iterate over IApiDescriptionGroupCollectionProvider.ApiDescriptionGroups) .

Thanks for the help!
